### ISSUE TITLE
Pyscript function task.wait_until() is blocking main thread

### ISSUE THREAD

**Issue by jojo141185** [custom-components/pyscript#150]
**Background**
In my PyScript code "CoverController" several objects of the class "Cover" are created. After a certain time an update routine of these objects is called, which should adjust the position of the cover depending on the current sun position.
```
     @time_trigger(f"cron(*/{COVER_UPDATE_MINUTES} * * * *)")
     def timebased_trigger():
       self.update_covers()
     registered_triggers.append(timebased_trigger)

    def update_covers(self):
        sun_elevation = float(state.get("sun.sun.elevation"))
        sun_azimuth = float(state.get("sun.sun.azimuth"))
        for cover in self.__covers:
            cover.update_cover(
                sun_elevation,
                sun_azimuth
            )
```
To adjust the cover position, the new position must always be approached first and after that, if necessary, the tilt position must be adjusted. To control the position and then the tilt-position sequentially I use the PyScript function task.wait_until(). It waits until the new position is reached, i.e. the cover-object leaves the state CLOSING or OPENING. 
```
    @task_unique('set_position')
    def set_position(self):
        current_position = state.get(
            self.__entity_id+".current_position"
        )
        if current_position != self.__position:
            self.set_cover_position()
            trig_info = task.wait_until(
                state_trigger=f"{self.__entity_id} != 'closing' and {self.__entity_id} != 'opening'", 
                timeout=250,
                state_check_now=False
            )
            if trig_info["trigger_type"] == "timeout":
                # timeout
                CoverController.log_error(self,f"Cover {self.__entity_id} is moving unnormal long. Timeout.")
            else:
                CoverController.log_debug(self,f"Cover {self.__entity_id} set to position: {self.__position}.")
        if self.__tilting:
            current_tilt_position = state.get(
                self.__entity_id+".current_tilt_position"
            )
            if current_tilt_position != self.__tilt_position:
                self.set_tilt_position()
                trig_info = task.wait_until(
                    state_trigger=f"{self.__entity_id} != 'closing' and {self.__entity_id} != 'opening'", 
                    timeout=250,
                    state_check_now=False
                )
                if trig_info["trigger_type"] == "timeout":
                    # timeout
                    CoverController.log_error(self,f"Cover {self.__entity_id} is tilting unnormal long. Timeout.")
                else:
                    CoverController.log_debug(self,f"Cover {self.__entity_id} set to tilt position: {self.__tilt_position}.")

```
**The following problem / incorrect behaviour appears:**
Not only the task execution within the function of the respective object is stopped, the whole flow of the script seems to be interrupted and waits until the position of the first cover-object has been approached, before updating the second cover, etc.

My goal would be that task.wait_until() only interrupts the process flow within the respective function/object and the different cover-objects are updated in parallel.
- Is there a way to implement this in PyScript? 
- Is this a desired behavior of task.wait_until()? I thought the pyscript functions run async.

**Comment by craigbarratt** [comment]
Pyscript can definitely handle this.  But there are various ways to split functions you need between different tasks.

Saying that pyscript functions are "async" doesn't mean each function automatically runs independently. A task is required to create an independent thread of execution.  Within a task, the code executes sequentially as usual. Multiple tasks run in parallel, independent of each other (unless you impose some requirements, eg: `task.unique`). When a trigger occurs it automatically creates a new task, so each trigger creates a new independent thread of execution.

It's not clear from your code where and how `set_position` is called.  If `update_cover` calls `set_position` then you have strictly sequential execution - every `COVER_UPDATE_MINUTES`, one new task is created that runs the function `timebased_trigger`.  That, in turn calls `update_covers`, which loops through each cover calling `update_cover`.  Assuming that calls `set_position`, which waits for each cover to achieve its correct position and then tilt.  Moreover, since you have a `@task_unique('set_position')` decorator on `set_position`, only one task can be running that function at any time (but you appear to only have a single task running that function, so that decorator will have no effect).

There are several solutions to make each cover act in parallel.  One approach would be to have one instance of each class per cover, with it's own time trigger.  So with `n` covers, you would have `n` instances of the class, and `n` time triggers would all fire at the same time, and would run `update_cover` and `set_position` on each cover independently.

An alternative approach which is likely simpler in terms of fewest changes to your code is to manually create a new task each time you run `set_position`.  You can do that with `task.create`.  So instead of calling:
```python
self.set_position()
```
instead you would call:
```python
task.create(self.set_position())
```
In both approaches you will need to remove the `@task_unique('set_position')` decorator, and instead call `task.unique` inside the function with a name that is unique to that cover, so that there is only one `set_position` task running per cover, eg:
```python
task.unique(f"set_position_{self.__entity_id}")
```

**Comment by jojo141185** [comment]
Okay, I got it now. Implemented as described and everything works :-) 
@craigbarratt Thank you so much for your clear and unambiguous explanations!
